types/prettier/index.d.ts

```diff
@@ -24,20 +29,160 @@ export type LiteralUnion<T extends U, U = string> = T | (Pick<U, never> & { _?:
 export type AST = any;
 export type Doc = doc.builders.Doc;
 
-// https://github.com/prettier/prettier/blob/main/src/common/ast-path.js
+// The type of elements that make up the given array T.
+type ArrayElement<T> = T extends Array<infer E> ? E : never;
+
+// A union of the properties of the given object that are arrays.
+type ArrayProperties<T> = { [K in keyof T]: T[K] extends any[] ? K : never }[keyof T];
+
+// A union of the properties of the given array T that can be used to index it.
+// If the array is a tuple, then that's going to be the explicit indices of the
+// array, otherwise it's going to just be number.
+type IndexProperties<T extends { length: number }> = IsTuple<T> extends true
+    ? Exclude<Partial<T>['length'], T['length']>
+    : number;
+
+// Effectively performing T[P], except that it's telling TypeScript that it's
+// safe to do this for tuples, arrays, or objects.
+type IndexValue<T, P> = T extends any[] ? (P extends number ? T[P] : never) : P extends keyof T ? T[P] : never;
```

Comment:
Why are you unable to use `T[P]`?

Reply:
Would love to, but cannot find a way to tell TypeScript that the `CallProperties` type means that `P` can index `T`. Feel free to mess around with this [playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggTnAhiACnA9pOwCWEDOAPACoB8UAvFAN5QDaA0lDgHZQDWEI6AZlMQF0AXP0YCoEAB7AILACb4oiFiDriA-FCYiWEAG4Q4UAL51O3PoIDcAWABQoSFACS8qWkyHcBEhOmyFGigAG1kAc2AACx0AVwBbACNDE3IqZ3xiGLBQknIpGXlFYDgY6E0AUUkAY2CYuQhCFERsHERg3LoAIlCWCMjOgQAaUW7wqIHyWMTDWwdwaFd6yQA1NtKSYZRU-j8CwOVVDSgUXYDFFniko01iOhRxHX1kkRP8s44uXh2bu4eoXQMcFm9kcCwyWRyZEoOzehXoR2KpSgImIpzhdFYPGSADEcHB8MBhgA6EmY5IAJQgcUQrCO6Uy2QalOprEmUB4bXwEGBcycAGE2sEPFhvEQoVRUbD9io1FBNIt3BgRXgxWzzF9iLNQS4ZHBhV4VbloZL-HCDrL5W5JPqWj4oSJ4EhUEqDXbSDztQLgsEvcEEogquwNlAAKrbAAUYEQUREAGVijgqsAmlFcsNWEspldhno1hARAcAJSUcghnmsXUcqrQeNwRPJ6ORI3UexQKBVQWEMPhjve-2BkS+339oPEYZhwsiMut9udkObACMaMCvptotypB7gpHg8Fw4DQYVKzzwZQC9I44vUDASoXLwXk9Dszbvfa8+OS6linVlk2ACZl0UVcXVtMU7gXARSE3V8dygIdtwPQgj1WWoGluM8hmOP8rzDYYb0wO8PzwpU-xeP9H2nOwXznRdAI+Cx+H-OjgM8UCSHAyDNgAZmYwU10NZCT3QiD-yg6CEIHOC9wkw8rRQ9ZBNQpC5KExcrxQbDuJwq98LAQiz2IzBSKwwywC4l4uIo59Z29LtaK-eiNSYhyWOVO0OPUniXL4kD10U9ZhMwjSoM2AAWXjvX4nx-IaGL2Iw0StKgrc+wPXdvX3QNlKWeTYpUpS4riscP3UzTji49TQu00z9IXUzjI00zzPK0zQpeUKrJnV87OOOi1xABguHE1LJPgkagwOS8apeeqZuvJVmpQVq5pJIldPwF41E6uxjHsIA).

We can always replace `IndexValue<T, P>` with `T[P]` if `P` explicitly extends `keyof T`. But if you're trying to filter the properties (like I am with `CallProperties`), I can't seem to get it to understand that it's returning a subset.